File: July2022/C++/PascalsTriangle.cpp
class Solution {
public:
    // TC: O(numRows^2)
    // SC: O(2 * numRows)
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> ans;
        vector<int> curr, prev;
        
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    curr.push_back(1);
                } else {
                    curr.push_back(prev[j - 1] + prev[j]);
                }
            }
            ans.push_back(curr);
            prev = curr;
            curr.clear();
        }
        
        return ans;
    }
};

File: July2022/Java/PascalsTriangle.java

// @saorav21994
// TC: O(n*(n-1)) = O(n^2)
// SC: O(n^2) - output

class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 1; i <= numRows; i++) {
            List<Integer> tmp = new ArrayList<Integer>();
            int n = 1;
            for (int j = 1; j <= i; j++) {
                tmp.add(n);
                n = n*(i-j)/j;
            }
            res.add(tmp);
        }
        return res;
    }
}
